apps/emqx_rule_engine/src/emqx_rule_monitor.erl

```diff
@@ -115,7 +119,13 @@ retry_loop(resource, ResId, Interval) ->
             end;
         not_found ->
             ok
-    end.
+    end;
+
+retry_loop(refresh, resources_and_rules, _) ->
+    %% NOTE: the order matters.
+    %% We should always refresh the resources first and then the rules.
+    ok = emqx_rule_engine:refresh_resources(),
+    ok = emqx_rule_engine:refresh_rules().
```

Comment:
what happens if the functions return something else other than `ok` ?

Reply:
the `EXIT` handle_info clause does not seem to have a different logic when handing `normal` or other reasons.
is it a best-effort thing? do we rely on user to trigger a manual retry of the failed resources ?
maybe add them as code comments in the `EXIT` handling clause.

Reply:
The two functions only return `ok` or failed (crash) if something goes wrong, although these functions are unlikely fail because they have try..catch clauses.

And yes, we do the best effort in the `emqx_rule_monitor` module, let the user to reconnect the resources and then enable the rule in the worst cases.

I will add some comments.